Kubeadm added unit tests for pkg app/util

dmmcquay opened the pull request:
Added unit tests for kubeadm/app/util package testing functionality of tokens.go, error.go, and kubeconfig.go.

This PR is part of the ongoing effort to add tests (#35025)

/cc @pires @jbeda 

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35231)

<!-- Reviewable:end -->

dmmcquay committed:
kubeadm: added unit tests for util/tokens

dmmcquay committed:
kubeadm: added tests for util/{error,kubeconfig}

dmmcquay commented:
/cc @pires

pires commented:
@k8s-bot ok to test

jbeda commented on cmd/kubeadm/app/util/error_test.go:
```diff
@@ -0,0 +1,51 @@
+	}
+
+	for _, rt := range tokenTest {
+		checkErr("", rt.e, errHandle)
```
You should set `codeReturned` to 0 at the top of this loop so that you can detect when errHandle *isn't* called.

jbeda commented on cmd/kubeadm/app/util/error_test.go:
```diff
@@ -0,0 +1,51 @@
+		expected int
+	}{
+		{nil, 0},
+		{fmt.Errorf(""), 1},
```
Perhaps refer to these by symbol (`DefaultErrorExitCode`) vs. hard coding here.

jbeda commented on cmd/kubeadm/app/util/kubeconfig_test.go:
```diff
@@ -0,0 +1,174 @@
+}
+
+func init() {
+	rand.Seed(time.Now().UTC().UnixNano())
```
Most tests don't seed Rand so that things are predictable from run to run.

jbeda commented on cmd/kubeadm/app/util/kubeconfig_test.go:
```diff
@@ -0,0 +1,174 @@
+	}
+	for _, rt := range createBasicTest {
+		c := CreateBasicClientConfig(rt.cc.c, rt.cc.s, rt.cc.ca)
+		if c.Kind != rt.expected {
```
This seems like a pretty weak test.  Is this just testing that things don't blow up?

dmmcquay replied:
Yeah, basically just checks so things don't blow up. Obviously more tests could get added here but wanted to at least get the structure checked.

jbeda commented on cmd/kubeadm/app/util/kubeconfig_test.go:
```diff
@@ -0,0 +1,174 @@
+	}
+}
+
+func setEnvs() {
```
There are a couple of problems here.  First, you can't assume that `/tmp` is a directory you can use.  Second, the random number may not be unique -- especially over time.  Third, you are cleaning up after yourself.

The pattern to use seems to be to use `ioutil.TempDir` and remote that directory after you are done.  See https://github.com/kubernetes/kubernetes/blob/master/pkg/client/unversioned/clientcmd/loader_test.go#L128.

Finally -- it is unfortunate that you are having to set the environment variables.  I'd be afraid those would leak places you wouldn't expect.  It might make more sense to modify `GetEnvParams` so that you inject some values there.

jbeda commented on cmd/kubeadm/app/util/kubeconfig_test.go:
```diff
@@ -0,0 +1,174 @@
+	for _, rt := range writeConfig {
+		c := CreateBasicClientConfig(rt.cc.c, rt.cc.s, rt.cc.ca)
+		err := WriteKubeconfigIfNotExists(rt.name, c)
+		if err != rt.expected {
```
It would be much stronger to verify what is written here.

jbeda commented on cmd/kubeadm/app/util/kubeconfig_test.go:
```diff
@@ -0,0 +1,174 @@
+		cc       configClient
+		expected error
+	}{
+		{fmt.Sprintf("%d", rand.Int()), configClient{}, nil},
```
Any reason why this name is random?  That will just make it harder to debug?

jbeda commented on cmd/kubeadm/app/util/tokens_test.go:
```diff
@@ -83,3 +84,87 @@ func newSecretsWithToken(token string) *kubeadmapi.Secrets {
+		givenToken := strings.Split(strings.ToLower(rt.s.GivenToken), ".")
+		if len(givenToken) != rt.l {
+			t.Errorf(
+				"failed GenerateToken:\n\texpected: %d\n\t  actual: %d",
```
Make these error cases more specific.  "failed GenerateToken num parts:"

jbeda commented on cmd/kubeadm/app/util/tokens_test.go:
```diff
@@ -83,3 +84,87 @@ func newSecretsWithToken(token string) *kubeadmapi.Secrets {
+		}
+		if len(givenToken[0]) != rt.n {
+			t.Errorf(
+				"failed GenerateToken:\n\texpected: %d\n\t  actual: %d",
```
Same here -- "failed GenerateToken first part length"

jbeda commented on cmd/kubeadm/app/util/tokens_test.go:
```diff
@@ -83,3 +84,87 @@ func newSecretsWithToken(token string) *kubeadmapi.Secrets {
+		{kubeadmapi.Secrets{GivenToken: "abcdef.a"}, true},
+	}
+
+	for _, rt := range tokenTest {
```
You should check the error here too.

pires replied:
Done.

jbeda commented on cmd/kubeadm/app/util/tokens_test.go:
```diff
@@ -83,3 +84,87 @@ func newSecretsWithToken(token string) *kubeadmapi.Secrets {
+		actual, _ := UseGivenTokenIfValid(&rt.s)
+		if actual != rt.expected {
+			t.Errorf(
+				"failed UseGivenTokenIfValid:\n\texpected: %t\n\t  actual: %t",
```
Making this error more descriptive (pulling out the given token, for example) will make this easier to debug if it ever fails.

dmmcquay committed:
kubeadm: EnvParam struct and GlobalEnvParam

Previously, GetEnvParams (now called SetEnvParams) had no way of being altered unless
it was through enviroment variables. These changes allow for a global
EnvParam to be set and also altered while still initally getting their value from
set enviroment variables. This change is especially helpful for testing
(see kubeadm/app/util/kubeconfig_test.go).

dmmcquay commented:
@jbeda I fixed most of the changes you asked for, let me know what you think!

jbeda commented on cmd/kubeadm/app/apis/kubeadm/env.go:
```diff
@@ -25,7 +25,7 @@ import (
 // TODO(phase2) use componentconfig
 // we need some params for testing etc, let's keep these hidden for now
-func GetEnvParams() map[string]string {
+func GetEnvParams(env map[string]string) map[string]string {
```
Passing a map in here is a bit awkward.  That means we have to pass a environment map down to anything that might use it.

At the very least, I'd do the following:
* Define a real type for the environment.  Either keep it a map `type EnvParams map[string]string]` or create a real struct in types.go.
* Have GetEnvParams just return one of those.  No need to pass anything in.

Then you can pass that more strongly typed and self documenting thing down.  Or we can keep it global but a bit cleaner with something like:

```
type EnvParams struct {
  KubernetesDir string
  ...
}

func GetEnvParams() *EnvParams {
...
}
```

You can then doctor it how you like and pass it down.  Or we can keep this as "global" but have it as an easy thing to swap in and out/duplicate.

```
var GlobalEnvParams = GetEnvParams()
```

And then in the test:
```
oldEnv := GlobalEnvParams
GlobalEnvParams = GetEnvParams()
GlobalEnvParams.foo = bar
defer func () { GlobalEnvParams = oldEnv }
```

dmmcquay replied:
When you say `types.go`, is that creating another file in the same pkg (kubeadm/app/apis/kubeadm) or some global k8s types file?

dmmcquay replied:
I took a stab at creating a GlobalEnvParams, hopefully I was able to capture what you mean.  Please let me know!

jbeda reviewed (changes requested):
Looking better!  The whole env thing is a little tricky.  Hard to figure out how to change that stuff without polluting a function signature.  Capturing the environment early (at start time) and being able to swap it out is cleaner and more isolated without having to modify a bunch of functions.

jbeda commented:
Hopefully you don't mind doing another iteration.  Sorry for not being clearer the first time.

dmmcquay commented:
Not a problem at all, I really appreciate it. I am learning a bunch.  Thanks!

jbeda commented:
Very cool.  Looks good to me.

pires commented:
@k8s-bot OK verify test this

pires commented:
I believe this is related to https://github.com/kubernetes/kubernetes/issues/35359.

jbeda commented:
The bazel stuff is disabled now. You can rebase and we should be good to go.  I'll re-LGTM after that.

dmmcquay commented:
@jbeda went ahead and did the rebase. Let me know if you want me to change anything else.

pires commented:
@jbeda it seems you still have to approve the changes you requested?

jbeda commented:
Looks like you need to run `hack/update-bazel.sh`

dmmcquay commented:
@jbeda ran hack/update-bazel.sh. Will that also need to be run for #35326 (haven't run hack/update-bazel.sh for that PR) and if so, should I just do that now?

jbeda commented:
@k8s-bot verify test this

Not sure why bazel verify is still failing.

mikedanese commented:
@dmmcquay the verify is failing because there is a merge conflict. Please rebase and try again.

dmmcquay commented:
@mikedanese I rebased, it was a conflict with a `BUILD` file.

dmmcquay commented:
@jbeda @mikedanese seems like hack/update-bazel.sh is failing, although I just ran it and it seems to pass. Any ideas?

mikedanese commented:
I can look tomorrow

mikedanese commented on cmd/kubeadm/app/node/BUILD:
```diff
@@ -33,3 +33,17 @@ go_library(
+
+go_test(
+    name = "go_default_test",
+    srcs = ["bootstrap_test.go"],
```
This file doesn't exist so presubmits are failing. Can you run ./hack/update-bazel.sh with a clean workspace e.g. git stash then update bazel then git stash pop?

Commit:
kubeadm: gofmt'ed cmd/kubeadm.

Commit:
kubeadm: updated cmd/kubeadm/app/node package related BUILD file.

pires commented:
I'm okay with these commits being contributed to Google.

@mikedanese done.

pires commented:
@jbeda care to fix CLA and lgtm? Thanks.

luxas commented:
I haven't had time to read this one thought fully but trusting earlier reviewers, it should be just fine.
Nits, if any, can be fixed in follow-ups.

We want to keep up the speed, so merging...

luxas commented:
@k8s-bot unit test this

luxas commented:
@k8s-bot unit test this
